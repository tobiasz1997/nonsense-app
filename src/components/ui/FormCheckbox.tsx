import cx from 'classnames';
import React, {
	forwardRef,
	ForwardRefRenderFunction,
	InputHTMLAttributes
} from 'react';
import FormErrorMessage from '@components/ui/FormErrorMessage';

type FormCheckboxProps = {
	label: string;
	error?: string;
	type?: 'checkbox' | 'radio';
};

const FormCheckbox: ForwardRefRenderFunction<
	HTMLInputElement,
	FormCheckboxProps & InputHTMLAttributes<HTMLInputElement>
> = (props, ref) => {
	return (
		<div>
			<label className={cx('flex w-full cursor-pointer text-green-dark')}>
				<span
					className={cx(
						'relative flex h-6 w-6 shrink-0 items-center justify-center'
					)}
				>
					<input
						{...props}
						ref={ref}
						type={props.type ?? 'checkbox'}
						className={cx(
							'peer h-full w-full cursor-pointer rounded border border-green-dark checked:border-green-dark checked:bg-none checked:text-white checked:hover:border-green-dark focus:ring-4 focus:ring-yellow checked:focus:border-green-dark'
						)}
					/>
					<span
						className={cx(
							'absolute left-[4px] h-4 w-4 rounded-sm bg-green-dark opacity-0 transition duration-200 peer-checked:opacity-100'
						)}
					/>
				</span>

				<span className="ml-3 flex items-center text-xs md:text-base md:leading-5">
					{props.label}
					{props.required && (
						<span className="my-0.5 mx-1 h-1 w-1 rounded-full bg-red" />
					)}
				</span>
			</label>
			{props.error && <FormErrorMessage>{props.error}</FormErrorMessage>}
		</div>
	);
};

export default forwardRef(FormCheckbox);
