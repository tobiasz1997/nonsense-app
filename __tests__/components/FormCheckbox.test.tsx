import { fireEvent, render, screen } from '@testing-library/react';
import FormCheckbox from '@components/ui/FormCheckbox';

describe('Form Checkbox', () => {
	const label = '@#$%^&*()qwerty';

	it('should render checkbox without crashing', () => {
		const { container } = render(<FormCheckbox label={label} />);
		expect(container).toMatchSnapshot();
	});

	it('should render radio without crashing', () => {
		const { container } = render(<FormCheckbox label={label} type="radio" />);
		expect(container).toMatchSnapshot();
	});

	it('should have correct value on click', async () => {
		render(<FormCheckbox label={label} />);

		const checkbox = screen.getByRole('checkbox');
		fireEvent.click(screen.getByText(label));

		expect(checkbox).toBeChecked();
	});

	it('should have correct value on double click', async () => {
		render(<FormCheckbox label={label} />);

		const checkbox = screen.getByRole('checkbox');
		fireEvent.dblClick(screen.getByText(label));

		expect(checkbox).not.toBeChecked();
	});

	it('should only one radio checkbox be checked', async () => {
		const labelOne = 'labelOne';
		const labelTwo = 'labelTwo';

		render(
			<form>
				<FormCheckbox
					label={labelOne}
					name="test"
					type="radio"
					data-testid={labelOne}
				/>
				<FormCheckbox
					label={labelTwo}
					name="test"
					type="radio"
					data-testid={labelTwo}
				/>
			</form>
		);

		const checkboxOne = screen.getByTestId(labelOne);
		const checkboxTwo = screen.getByTestId(labelTwo);
		fireEvent.click(screen.getByTestId(labelOne));
		fireEvent.click(screen.getByTestId(labelTwo));
		fireEvent.click(screen.getByTestId(labelOne));

		expect(checkboxOne).toBeChecked();
		expect(checkboxTwo).not.toBeChecked();
	});
});
