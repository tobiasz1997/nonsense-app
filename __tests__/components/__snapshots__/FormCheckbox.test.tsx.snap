// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Checkbox should render checkbox without crashing 1`] = `
<div>
  <div>
    <label
      class="flex w-full cursor-pointer text-green-dark dark:text-pistachio"
    >
      <span
        class="relative flex h-6 w-6 shrink-0 items-center justify-center"
      >
        <input
          class="peer h-full w-full cursor-pointer border border-green-dark checked:border-green-dark checked:bg-none checked:text-white checked:hover:border-green-dark focus:ring-4 focus:ring-yellow checked:focus:border-green-dark disabled:opacity-70 rounded"
          label="@#$%^&*()qwerty"
          type="checkbox"
        />
        <span
          class="absolute left-[4px] h-4 w-4 bg-green-dark opacity-0 transition duration-200 peer-checked:opacity-100 rounded-sm"
        />
      </span>
      <span
        class="ml-3 flex items-center text-xs md:text-base md:leading-5"
      >
        @#$%^&*()qwerty
      </span>
    </label>
  </div>
</div>
`;

exports[`Form Checkbox should render radio without crashing 1`] = `
<div>
  <div>
    <label
      class="flex w-full cursor-pointer text-green-dark dark:text-pistachio"
    >
      <span
        class="relative flex h-6 w-6 shrink-0 items-center justify-center"
      >
        <input
          class="peer h-full w-full cursor-pointer border border-green-dark checked:border-green-dark checked:bg-none checked:text-white checked:hover:border-green-dark focus:ring-4 focus:ring-yellow checked:focus:border-green-dark disabled:opacity-70 rounded-full"
          label="@#$%^&*()qwerty"
          type="radio"
        />
        <span
          class="absolute left-[4px] h-4 w-4 bg-green-dark opacity-0 transition duration-200 peer-checked:opacity-100 rounded-full"
        />
      </span>
      <span
        class="ml-3 flex items-center text-xs md:text-base md:leading-5"
      >
        @#$%^&*()qwerty
      </span>
    </label>
  </div>
</div>
`;
